proxyservice/gcr.go

```diff
@@ -0,0 +1,125 @@
+package proxyservice
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+
+	"github.com/docker/distribution/reference"
+)
+
+// pubSubNotification is the JSON message Google PubSub sends to our HTTP
+// endpoint
+type pubSubNotification struct {
+	Message      pubSubMessage
+	Subscription string
+}
+
+type pubSubMessage struct {
+	Attributes map[string]string
+	// Base64-ed JSON GcrWebhookData object
+	Data      string
+	MessageId string
+}
+
+// GcrWebhookData is the GCR-related data
+type GcrWebhookData struct {
+	Action string `json:"action"`
+	Digest string `json:"digest,omitempty"`
+	Tag    string `json:"tag,omitempty"`
+}
+
+func (d *GcrWebhookData) hasValidAction() bool {
+	validActions := map[string]bool{"INSERT": true}
+	return validActions[d.Action]
+}
+
+func (d *GcrWebhookData) hasValidImageReference() bool {
+	return reference.ReferenceRegexp.MatchString(d.getImageReference())
+}
+
+func (d *GcrWebhookData) isValid() bool {
+	if !d.hasValidAction() {
+		return false
+	}
+	if !d.hasValidImageReference() {
+		return false
+	}
+	return true
+}
+
+func (d *GcrWebhookData) getImageReference() string {
+	if d.Tag != "" {
+		return d.Tag
+	} else {
+		return d.Digest
+	}
+}
+
+func (d *GcrWebhookData) getImageTagOrDigest() string {
+	if d.Tag != "" {
+		return reference.ReferenceRegexp.FindStringSubmatch(d.Tag)[2]
+	} else {
+		return reference.ReferenceRegexp.FindStringSubmatch(d.Digest)[3]
+	}
```

Comment:
Hmmm, I don't think so, unless Google messes up. In any case, the out of range read should only crash the handling goroutine, right? We could definitely protect against it, though.

Reply:
I think it makes sense to check the length and return an error if it's the wrong length or if the regex doesn't match.

Reply:
the length of the slice returned by `FindStringSubmatch` is based on the submatch/capture groups in the regexp. so the regexp that we import would have to change for this to be _broken_. it could return an empty string, though.

Reply:
If `FindStringSubmatch` doesn't match, it will return `[]`, so we should just check that the length of the returned list is >= 3 and 4.